pep-9999.rst

```diff
@@ -0,0 +1,173 @@
+
+- ``_next_in_mro`` speed hack will be not necessary since subscription will
+  not create new classes.
+
+- Ugly ``sys._getframe`` hack, this one is particularly nasty, since it looks
+  like we can't remove it without changes outside ``typing``.
+
+- Currently generics do dangerous things with private ABC caches
+  to fix large memory consumption that grows at least as O(N\ :sup:`2`),
+  see [2]_. This point is also important because it was recently proposed to
+  re-implement ``ABCMeta`` in C.
+
+- Problems with sharing attributes between subscripted generics,
+  see [3]_. Current solution already uses ``__getattr__`` and ``__setattr__``,
+  but it is still incomplete, and solving this without the current proposal
+  will be hard and will need ``__getattribute__``.
+
+- ``_no_slots_copy`` hack, where we clean-up the class dictionary on every
+  subscription thus allowing generics with ``__slots__``.
+
+- General complexity of the ``typing`` module, the new proposal will not
+  only allow to remove the above mentioned hacks/bugs, but also simplify
+  the implementation, so that it will be easier to maintain.
+
+
+Specification
+=============
+
+Idea of ``__class_getitem__`` is very simple, it is an exact analog of
+``__getitem__`` with an exception that it is called on a class that
+defines it, not on its instances, this allows us to avoid
+``GenericMeta.__getitem__`` for things like ``Iterable[int]``. Note that this
+method is used as a fallback, so if a metaclass defines ``__getitem__``,
+then it will have the priority.
+
+If an object that is not a class object appears in bases of a class
+definition, the ``__subclass_base__`` is searched on it. If found,
+it is given an original tuple of bases as an argument. If the result of call
+is not ``None``, then it is substituted instead of this object.
+Otherwise, the base is just removed. This is necessary to avoid
+inconsistent MRO errors, that are currently prevented by manipulations in
+``GnericMeta.__new__``. After creating the class, original bases are saved
+in ``__orig_bases__`` (now this is also done by the metaclass).
+
+These two method names are reserved for exclusive use by the ``typing`` module
+and the generic types machinery, and any other use is strongly discouraged.
+The reference implementation (with tests) can be found in [4]_, the proposal
+was originally posted and discussed on the ``typing`` tracker, see [5]_.
+
+
+Backwards compatibility and impact on users who don't use ``typing``:
+=====================================================================
+
+This proposal will break code that currently uses the names
+``__class_getitem__`` and ``__subclass_base__``.
+
+This proposal will allow to have almost full backwards compatibility with
+current public generic types API, moreover the ``typing`` module is still
+provisional.
+
+With the reference implementation I measured negligible performance effects
+(under 1%) for regular (non-generic) classes.
```

Comment:
Would like to see how you measured this -- in a microbenchmark or for some app? And have you found perf improvements for generics?

Reply:
Yes, I tested it on a micro-benchmark, here is the code:
```python
class A: ...
class B: ...

for i in range(100000):
    class E(A, B):
        cx = 0
        cy = 0
        def __init__(self, x, y):
            self.x = x
            self.y = y
```
The difference is statistically insignificant:
* master: 3.19 +- 0.04 s
* proposed: 3.15 +- 0.04 s

It is hard to measure speed-up for generics in ``typing`` without writing the full new ``typing`` implementation, but here are some basic observations:
* The first ``List[int]`` will be at least 10x faster.
* All next appearances of ``List[int]`` will take the same time, since type cache look-up is 2x faster than the new ``__class_getitem__``.
* Method calls on instances of generic classes may be up to 10% faster.
* Instantiating generic classes probably will be few percent faster, but it is hard to say for sure now.

Also there will be a decrease in memory consumption. It is hard to say how much, but my expectations are that ``typing`` will be around 2x "lighter". Also annotations in user code will save some memory (although again this is not dramatic due to generic type cache).

To summarize, the speed/memory improvements are rather incremental (although quite good), this is why I have two additional motivation points: avoiding metaclass conflicts and the removal of various hacks.